pytorch_lightning/loggers/comet.py

```diff
@@ -114,8 +116,17 @@ def __init__(self,
             self.mode = "offline"
             self.save_dir = save_dir
         else:
-            # If neither api_key nor save_dir are passed as arguments, raise an exception
-            raise MisconfigurationException("CometLogger requires either api_key or save_dir during initialization.")
+            # for backwards-compatibility, we have to check api_key then save_dir and
+            # only then use ~/.comet.config or the environment variable
+            api_key = get_api_key(None, get_config())
```

Comment:
I think this can be checked above before `if api_key is not None:` to avoid duplicate code below.
```
if api_key is None:
    api_key = get_api_key(None, get_config())
```

Reply:
I was trying to preserve the previous behavior of 

1. If you pass in an API key, it is used (regardless of `save_dir`)
2. If no API key is given but `save_dir` is, an offline experiment is used

Step 3. before would have been to error; now step 3 is to use the API key from the environment / the config file if given and only error if not. If we load the API key in at the start, then behavior 2. can't be done anymore (i.e. if I have a `~/.comet_config` but still want the experiment to be saved locally, I could pass in `save_dir` and that would be used instead). I like keeping explicit local saving as an option, so if we're going to change the behavior, I'd suggest

1. if `save_dir` is given, do an offline experiment (regardless of API key status; breaks backwards-compatibility)
2. if not, use the given API key if there is one or load it from the env variable / file if not
3. if none of these work, throw the error.

Reply:
Yeah! the suggested workflow looks like a better way to make this work, although the current one is also right but not optimal(just repeated code).

Reply:
What do you think about throwing an error if `save_dir` and `api_key` are given? This also breaks backwards-compatibility but not silently. By disallowing this, we remove potential confusion about which argument has precedence if both are given. And then we can proceed with the second option above.

Reply:
Yeah, I was thinking the same. For now, api_key is given priority if both are given. As a user, I don't think I will bother to give a save_dir even if I want to run it in online mode. I will give save_dir only when I want to run it in offline mode even if I have a key already present in my system or I will give an api_key. So I think the suggested flow looks better to me atleast but you should talk to any of the core-contributor first before changing it.

Reply:
`save_dir` actually has another effect as well - setting it will change where the model weights are saved. I just ran into a case where I was doing offline experiments and using `save_dir` and then wanted to switch to online ones, but I think it's more inconvenient and error-prone to have to change the save directory somewhere else when I already have `save_dir` set up for that. Thus I think it actually could make sense to specify `save_dir` even if you want online-mode. What about adding a another argument to specify whether the experiment should be online?

* If `save_dir` is given and no API key can be found, we can infer it should be offline.
* If `save_dir` is not given, then it's online.
* If desired, explicitly giving an API key could lead to online mode.
* If an API key can be found (e.g. in `~/.comet.config`) but `save_dir` is also given, then we use the new argument to determine whether to be online or not. This lets users use `save_dir` to control where checkpoints are stored even when doing online experiments and lets users do offline experiments despite having a `~/.comet.config`.

We'd also want to change it so that `save_dir` is always saved as an instance variable (instead of only if no API key is given).

Reply:
For now, if `save_dir` is `None` then in case of online more where are models saved?? I think if someone wants to explicitly specify the model weights directory, one should use ModelCheckpoint rather than specifying the model directory in logger instance.